File: 백준/Silver/2210. 숫자판 점프/숫자판 점프.py
import sys
input = sys.stdin.read

board = [list(map(int, line.split())) for line in input().strip().splitlines()]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

unique_numbers = set()

def dfs(x, y, number):
    if len(number) == 6:
        unique_numbers.add(number)
        return

    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if 0 <= nx < 5 and 0 <= ny < 5:
            dfs(nx, ny, number + str(board[nx][ny]))

for i in range(5):
    for j in range(5):
        dfs(i, j, str(board[i][j]))

print(len(unique_numbers))

File: 백준/Silver/2210. 숫자판 점프/숫자판 점프.java
import java.io.*;
import java.util.*;

public class Main {
    static int[][] board = new int[5][5];
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static Set<String> uniqueNumbers = new HashSet<>();
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        for (int i = 0; i < 5; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 5; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                dfs(i, j, String.valueOf(board[i][j]));
            }
        }
        
        System.out.println(uniqueNumbers.size());
        
        br.close();
    }
    
    static void dfs(int x, int y, String number) {
        if (number.length() == 6) {
            uniqueNumbers.add(number);
            return;
        }
        
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if (nx >= 0 && nx < 5 && ny >= 0 && ny < 5) {
                dfs(nx, ny, number + board[nx][ny]);
            }
        }
    }
}